# Assignment 3: Selecting OPC Data For Review

## DSCI 504: Intro to SQL

Jeffery Boczkaja

05/25/23

**Step 1: Use a SELECT query to select all records from the customer table. Limit the displayed results to 10 records.**

**Q: Of the first ten customers returned, what is the Customer ID  and City of the 5th customer?**

```
A: cus_id 2604, cus_city Pataskala
```

In [1]:
SELECT *
FROM customers
LIMIT 10

(10 row(s) affected)

Total execution time: 00:00:00.016

cus_id cus_num cus_last_name cus_first_name cus_add_num cus_address cus_city cus_state cus_zip cus_phone cus_join_date cus_app_cd cus_app_num tot_ord_qty tot_ord_value 2600 NULL Holzer Yeni 8617 High Noon Street Hattiesburg 17 39401 7833472627 2020-11-22 5 OPC85554622 60 10641.48 2601 NULL Hull Tancredo 7673 Cedat Camp Road Oswego 19 13126 5154255063 2020-04-19 9 OPC86688258 49 24585.61 2602 NULL Sokol Teodor 965 Southampton Salt Lake City 26 84119 8802052764 2020-04-15 14 OPC87710870 48 8514.29 2603 NULL Barnes Theo 270 High Noon Ave Coram 19 11727 2779494240 2020-09-14 18 OPC78993136 76 19173.06 2604 NULL Tempest Simone 751 East Shadow Brook Dr. Pataskala 20 43062 7692163675 2020-06-24 11 OPC86812641 45 27325.72 2605 NULL Bullock Pankaj 7524 Van Dyke St. Southampton 22 18966 2812450867 2020-03-20 3 OPC85939415 11 1214.88 2606 NULL Ware Reina 656 Grand Ave Los Angeles 4 90008 3902290668 2020-09-21 6 OPC86472716 64 6718.85 2607 NULL Ray Rosalin 8564 Old Woodland Rd Cranston 23 2920 7944432375 2020-06-06 9 OPC87091916 43 14012.58 2608 NULL Stark Herb 7291 Race St. Ithaca 19 14850 2978433949 2020-06-27 5 OPC87321099 28 4054.52 2609 NULL Bohme Bohuslav 135 Homewood Ln Merrimack 19 3054 3147113467 2020-12-12 19 OPC86853764 96 15178.38

**Step 2: Use a SELECT query to select the Customer ID, Last Name, State, and Customer Appreciation Code of all customers who joined _from_ September 11, 2001 _through_ October 1, 2001**

**Q: How many customers joined in teh period from September 11, 2001 throught October 1, 2001?**

```
A: 8 Customers

```

In [2]:
SELECT cus_id
    , cus_last_name
    , cus_state
    , cus_app_num
FROM customers
WHERE cus_join_date BETWEEN '2001-09-11' AND
'2001-10-01'

(8 row(s) affected)

Total execution time: 00:00:00

cus_id cus_last_name cus_state cus_app_num 356 Delacruz 2 OPC80718306 464 Sherman 3 OPC85270843 495 Delacruz 18 OPC79747551 1587 Applewood 17 OPC84950666 1907 Acosta 29 OPC82133096 1936 Manchester 11 OPC84539106 2141 Brooks 19 OPC84407421 2194 Moore 28 OPC81668931

**Part 3: Looking at the same query from Part 2, write a SELECT query that returns all information for the two customers with the last name Delacruz**

**Q: What are three unique features about these two customers?**

```
A: Their cus_ids are 356 and 495. Their cus_app_num are OPC80718306 and OPC79747551. Their cus_add_num are 63740 and 53556.

```

In [3]:
SELECT *
FROM customers
WHERE cus_join_date BETWEEN '2001-09-11' AND
'2001-10-01' AND
cus_last_name = 'Delacruz'

(2 row(s) affected)

Total execution time: 00:00:00.008

cus_id cus_num cus_last_name cus_first_name cus_add_num cus_address cus_city cus_state cus_zip cus_phone cus_join_date cus_app_cd cus_app_num tot_ord_qty tot_ord_value 356 Del73426 Delacruz Margo 63740 Crystal Spring Corners Mcarthur Estates 2 89443 3875686732 2001-09-25 6 OPC80718306 70 34773 495 Del44118 Delacruz Brandon 53556 Apache Wolfs Mill 18 18151 7593410556 2001-09-23 1 OPC79747551 85 2396

**Part 4: Write a query to return all order data for the two Delacruz customers above. Be sure to include minimal person information to assist in identifying which person placed which order. You may need to use a simple join to get both the customer and order data. A template for that is below. Use the following data points in your query:**

- **Customer ID**
- **Customer Last Name**
- **Customer First Name**
- **Total Order Quantity**
- **Total Order Value**
- **All Order Data**

```
SELECT [table1.column], [table2.column]..
FROM [table1], [table2]
WHERE [conditions]

```

In [4]:
SELECT customers.cus_id
    , cus_last_name
    , cus_first_name
    , ord_id	
    , ord_date	
    , ord_tax_loc	
    , order_tot	
    , ord_ship_add	
    , ord_ship_date	
    , ord_track_num	
    , prod_id	
    , warehouse_id 
FROM customers 
JOIN orders ON orders.cus_id=customers.cus_id
WHERE cus_join_date BETWEEN '2001-09-11' AND
'2001-10-01' AND
cus_last_name = 'Delacruz'


(2 row(s) affected)

Total execution time: 00:00:00.008

cus_id cus_last_name cus_first_name ord_id ord_date ord_tax_loc order_tot ord_ship_add ord_ship_date ord_track_num prod_id warehouse_id 356 Delacruz Margo 2614702 2002-08-21 2 5200 FALSE 2002-08-25 PAC7774446-3 42 3 495 Delacruz Brandon 2051970 2005-07-28 18 5631.35 FALSE 2005-07-30 PAC2382180-2 48 1

**Q: What is one interesting fact revealed in the addition of the new data?**

```
A: They both paid similar amounts of money for their orders.

```

**Q: Explain how the addition of additional data aided in your assessment.**

```
A: I coulnd't see how much they had spent with only the customer data but by joining it with the orders table I could see how much each customer paid for their product.

```

**Part 5: Some of the most unique features mountain bikers seek out in their bikes are wheel sets, shocks, forks, and dropper posts. These items make or break a top- notch mountain bike. You know that part of the identification of whether or not OPC should break into the luxury MTB market is that they can build and market competitive bikes. Use a SELECT query to determine how many unique type of components there are in the components table. Specifically query for forks and dropper posts. Order your results by component category and component name in alphabetical order.**

**Q: How many unique combination of fork and dropper post were there in the data?**

```
A: 29 Droppers * 56 Forks = 1624 combinations

```

In [5]:
SELECT *
FROM components
WHERE comp_cat='fork' 
OR comp_cat='dropper'
ORDER BY comp_cat, comp_name ASC

(85 row(s) affected)

Total execution time: 00:00:00.007

comp_id comp_name comp_cost comp_supplier comp_cat 260 Bike Yoke Devine Stealth 168 37 dropper 237 Bike Yoke Revive 315 37 dropper 258 BrandX Ascend 125 42 dropper 238 Cannondale 2 89 17 dropper 245 Cannondale DownLow 89 17 dropper 262 Cube Dropper 129 43 dropper 236 DropWH18 78 36 dropper 247 Fox Transfer 180 28 dropper 259 Fox Transfer Factory 329 28 dropper 234 FusionX Manic 189 34 dropper 243 Giant Contact Switch 189 1 dropper 254 GT Dropkick 74 24 dropper 252 Ibis 90 22 dropper 241 KS Lev 123 38 dropper 248 KS Lev Integra 156 38 dropper 256 KS Lev SI 134 38 dropper 250 KS Rage-i 135 38 dropper 249 Niner RDO 114 3 dropper 244 OC2 78 61 dropper 246 OC2 Carbon 199 61 dropper 253 OneUp V2 78 39 dropper 242 OnOff Pija 100 23 dropper 235 Recon Dropper 113 35 dropper 255 Recon Elite Dropper 113 35 dropper 239 Ride Alpha 110 2 dropper 240 RockShox Reverb Steakth 349 30 dropper 257 SDG Tellis 250 40 dropper 261 TransX 126.9 13 dropper 251 X-Fusion Manix 199.96 73 dropper 115 DVO Onyx D1 160 785 29 fork 138 DVO Onyx D2 170 Air 790 29 fork 136 Fox 32 Factory 100 489 28 fork 142 Fox 32 Float Perf 100 Air 510 28 fork 164 Fox 34 Float Factory Grip 2 120 969 28 fork 150 Fox 34 Float Factory Grip 2 140 789 28 fork 140 Fox 34 Float Factory Grip 2 140 Air 1039 28 fork 128 Fox 34 Float Perf 120 429 28 fork 153 Fox 34 Float Perf 140 410 28 fork 162 Fox 34 Float Perf 140 Air 429 28 fork 141 Fox 34 Perf 130 429 28 fork 147 Fox 34 Perf 130 Air 429 28 fork 113 Fox 34 Rhythm 120 429 28 fork 135 Fox 36 Factory 150 Air 1069 28 fork 134 Fox 36 Factory 160 Air 1069 28 fork 155 Fox 36 Float Factory 160 Air 605 28 fork 163 Fox 36 Float Factory 170 Air 999 28 fork 124 Fox 36 Float Perf 140 Air 949 28 fork 144 Fox 36 Float Perf 150 880 28 fork 118 Fox 36 Float Perf 160 Air 849 28 fork 143 Fox 36 Perf 160 Air 840 28 fork 158 Fox 36 Perf 1650 Air 893 28 fork 125 Fox 36 Perf 170 Air 798 28 fork 165 Fox 36 Perf Elite 140 Air 789 28 fork 146 Fox 36 Perf Elite 150 Air 899 28 fork 166 Fox 36 Perf Elite 160 Air 789 28 fork 148 Fox 36 Perf Elite 180 Air 1024 28 fork 160 Fox 36 Rhythm 160 Air 556 28 fork 116 Fox 36 Rhythm Fit 3 150 849 28 fork 114 Fox 38 Factory Grip 2 180 Air 1249 28 fork 123 Fox 38 Factory Grip2 170 Air 1249 28 fork 132 Fox 38 Float Perf 170 Air 1020 28 fork 119 Fox 38 Grip 2 170 Air 989 28 fork 137 Lefty Ocho 100 Air 999 33 fork 126 Marzocchi Bomber Z2 130 Air 610 32 fork 159 Marzocchi Bomber Z2 Rail 120 Air 519 32 fork 121 Ohlins RXF36 M.2 170 Air 1149.97 31 fork 139 RockShox 35 Gold RL 140 Air 458 30 fork 151 RockShox 35 Gold RL 150 Air 458 30 fork 161 RockShox 35 Gold RL 160 Air 458 30 fork 167 RockShox Lyric Sel 150 Air 765 30 fork 130 RockShox Lyric Ult 150 Air 968 30 fork 122 RockShox Lyric Ult RC2 160 968 30 fork 120 RockShox Lyric Ult RC2 160 Air 968 30 fork 149 RockShox Lyric Ult RC2 170 975 30 fork 131 RockShox Pike Sel RC 140 713 30 fork 152 RockShox Pike Sel+ 140 Air 713 30 fork 157 RockShox Pike Ult 150 Air 917 30 fork 145 RockShox Pike Ult RC 130 Air 917 30 fork 117 RockShox Recon RL 120 274 30 fork 133 RockShox Recon RL 120 Air 255 30 fork 127 RockShox Recon RL 130 275 30 fork 129 RockShox Recon RL 130 Air 270 30 fork 168 RockShox YariRC 160 509 30 fork 154 RockShox YarRC 180 Air 509 30 fork 156 RockShox ZEB Ult 170 Air 915 30 fork

**Part 6: Write a new query to identify all builds that the fork in row 60 is included on in the data. Include the accompanying build data in your query. Use a conditional query and search for the component name in the query.**

**Q: What happened in the query?**

```
A: All the data entries had their comp_fork replaces with 123 so over 1000 results were pulled by the query.

```

In [6]:
SELECT *
FROM builds
, components
WHERE comp_fork = '123'
ORDER BY comp_name

(1084 row(s) affected)

Total execution time: 00:00:00.128

build_id comp_hbar comp_brake comp_wheels comp_fork comp_shock comp_shifter comp_cassette comp_derail comp_dropper comp_crank comp_chain_ring build_name comp_id comp_name comp_cost comp_supplier comp_cat 42 30 67 88 123 184 61 219 273 260 273 273 electric 172 NULL NULL shock 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 172 NULL NULL shock 71 43 67 91 123 184 61 219 273 260 273 273 precision 172 NULL NULL shock 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 172 NULL NULL shock 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 260 Bike Yoke Devine Stealth 168 37 dropper 71 43 67 91 123 184 61 219 273 260 273 273 precision 260 Bike Yoke Devine Stealth 168 37 dropper 42 30 67 88 123 184 61 219 273 260 273 273 electric 260 Bike Yoke Devine Stealth 168 37 dropper 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 260 Bike Yoke Devine Stealth 168 37 dropper 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 237 Bike Yoke Revive 315 37 dropper 71 43 67 91 123 184 61 219 273 260 273 273 precision 237 Bike Yoke Revive 315 37 dropper 42 30 67 88 123 184 61 219 273 260 273 273 electric 237 Bike Yoke Revive 315 37 dropper 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 237 Bike Yoke Revive 315 37 dropper 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 258 BrandX Ascend 125 42 dropper 71 43 67 91 123 184 61 219 273 260 273 273 precision 258 BrandX Ascend 125 42 dropper 42 30 67 88 123 184 61 219 273 260 273 273 electric 258 BrandX Ascend 125 42 dropper 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 258 BrandX Ascend 125 42 dropper 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 45 Burgtec Alloy 106 27 handlebar 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 45 Burgtec Alloy 106 27 handlebar 71 43 67 91 123 184 61 219 273 260 273 273 precision 45 Burgtec Alloy 106 27 handlebar 42 30 67 88 123 184 61 219 273 260 273 273 electric 45 Burgtec Alloy 106 27 handlebar 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 238 Cannondale 2 89 17 dropper 71 43 67 91 123 184 61 219 273 260 273 273 precision 238 Cannondale 2 89 17 dropper 42 30 67 88 123 184 61 219 273 260 273 273 electric 238 Cannondale 2 89 17 dropper 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 238 Cannondale 2 89 17 dropper 71 43 67 91 123 184 61 219 273 260 273 273 precision 42 Cannondale 2 Flat 50 17 handlebar 42 30 67 88 123 184 61 219 273 260 273 273 electric 42 Cannondale 2 Flat 50 17 handlebar 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 42 Cannondale 2 Flat 50 17 handlebar 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 42 Cannondale 2 Flat 50 17 handlebar 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 18 Cannondale 3 Riser 128 17 handlebar 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 18 Cannondale 3 Riser 128 17 handlebar 71 43 67 91 123 184 61 219 273 260 273 273 precision 18 Cannondale 3 Riser 128 17 handlebar 42 30 67 88 123 184 61 219 273 260 273 273 electric 18 Cannondale 3 Riser 128 17 handlebar 71 43 67 91 123 184 61 219 273 260 273 273 precision 245 Cannondale DownLow 89 17 dropper 42 30 67 88 123 184 61 219 273 260 273 273 electric 245 Cannondale DownLow 89 17 dropper 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 245 Cannondale DownLow 89 17 dropper 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 245 Cannondale DownLow 89 17 dropper 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 80 Crankbrothers Synth EN 599.98 50 wheel 71 43 67 91 123 184 61 219 273 260 273 273 precision 80 Crankbrothers Synth EN 599.98 50 wheel 42 30 67 88 123 184 61 219 273 260 273 273 electric 80 Crankbrothers Synth EN 599.98 50 wheel 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 80 Crankbrothers Synth EN 599.98 50 wheel 42 30 67 88 123 184 61 219 273 260 273 273 electric 262 Cube Dropper 129 43 dropper 71 43 67 91 123 184 61 219 273 260 273 273 precision 262 Cube Dropper 129 43 dropper 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce 262 Cube Dropper 129

**Part 7: Use the simple join concept from Part 4 to complete the query by joining the componet and build tables. Return all Build columns in your query, but not the component data.**

**Q: How many builds is the 'Fox 38 Factory Grip2 170 Air' included in?**

```
A: 4 builds

```

In [7]:
SELECT builds.build_id	
, comp_hbar	
, comp_brake	
, comp_wheels	
, comp_fork	
, comp_shock	
, comp_shifter	
, comp_cassette	
, comp_derail	
, comp_dropper	
, comp_crank	
, comp_chain_ring	
, build_name
FROM builds,
components
WHERE components.comp_id='123' AND builds.comp_fork='123'



(4 row(s) affected)

Total execution time: 00:00:00.008

build_id comp_hbar comp_brake comp_wheels comp_fork comp_shock comp_shifter comp_cassette comp_derail comp_dropper comp_crank comp_chain_ring build_name 71 43 67 91 123 184 61 219 273 260 273 273 precision 42 30 67 88 123 184 61 219 273 260 273 273 electric 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 57 6 56 85 123 188 279 223 279 237 269 279 strikeforce

**Part 8: Write a query to find the build names of the four builds from Part 7 above.**

In [8]:
SELECT build_name
FROM builds,
components
WHERE components.comp_id='123' AND builds.comp_fork='123'
ORDER BY build_name ASC

(4 row(s) affected)

Total execution time: 00:00:00.023

build_name electric megalodon precision strikeforce

**Part 9: Identify the most expensive product of 2019. Since we haven't learned aggregate functions yet, please do not use the max() function. Use conditional statements to return a single row with the product with the highest order value**

**Q: What is the product name of the product and the price of the mountain bike with the highest prive ordered in 2019?**

```
A: The product name is SB150 and the highest price is $8000

```

In [9]:
SELECT prod_name
    , prod_price
FROM products
ORDER BY prod_price DESC
LIMIT 10

(10 row(s) affected)

Total execution time: 00:00:00

prod_name prod_price SB150 8000 SB160 7699.99 E-Series 7565.66 Rail 7399 Wreckoning 7399 Megatower 6479.99 Gnarvana 6395 SuperFoxy 6240.99 Firebird 5899 Altitude 5899

**Part 10: Execute three additional queries on tables you haven't used yet. These can be at your discretion.**

In [10]:
SELECT prod_class
    , prod_name
    , prod_price
FROM products
ORDER BY prod_price DESC
LIMIT 30

(30 row(s) affected)

Total execution time: 00:00:00

prod_class prod_name prod_price Expert SB150 8000 Expert SB160 7699.99 Expert E-Series 7565.66 Expert Wreckoning 7399 Expert Rail 7399 Expert Megatower 6479.99 Expert Gnarvana 6395 Expert SuperFoxy 6240.99 Expert Altitude 5899 Expert Firebird 5899 Expert SB140 5800 Expert Following 5799 Expert G-Series 5656.41 Expert Stereo 5609.99 Expert SB130 5500 Expert X Trance 5499 Expert Shredd Dog 5375.99 Expert Trail Pistol 5375.99 Expert Stumpjumper 5295.9 Expert Enduro 5271.99 Expert Mojo 5199 Expert SB100 5099 Mid Onset 5000 Mid Element 4854.99 Mid Process 134 4750.99 Mid The Smash 4745 Mid SB115 4700 Mid AMD 4679 Mid Meta Am 4645.99 Mid Mega 4504.99

In [11]:
SELECT comp_cost
    , comp_cat
FROM components
ORDER BY comp_cost DESC
LIMIT 50

(50 row(s) affected)

Total execution time: 00:00:00.015

comp_cost comp_cat NULL shock 2199 wheel 1721.9 wheel 1450 wheel 1350 wheel 1249 fork 1249 fork 1149.97 fork 1069 fork 1069 fork 1039 fork 1024 fork 1020 fork 1007 wheel 999 fork 999 fork 995 wheel 989 fork 975 fork 969 fork 968 fork 968 fork 968 fork 949 fork 917 fork 917 fork 915 fork 899 fork 893 fork 880 fork 849 fork 849 fork 840 fork 798 fork 790 fork 789 fork 789 fork 789 fork 785 fork 765 wheel 765 fork 765 shock 713 fork 713 fork 699 shock 699 shock 669.99 shock 649.99 shock 610 fork 605 wheel

In [12]:
SELECT tax_rate
    , tax_location
FROM taxes
ORDER BY tax_rate DESC

(31 row(s) affected)

Total execution time: 00:00:00.008

tax_rate tax_location 8.875 MN 8.52 KS 7.25 NY 7.25 CA 7.25 CO 7 RI 7 IN 6.5 WA 6.35 CT 6 SC 6 FL 6 WV 6 IA 6 KY 6 TN 5.75 OH 5.6 AZ 5.5 NE 5.5 ME 4.85 UT 4.75 NC 4.45 LA 4.3 VA 4 AL 4 HI 4 WY 4 GA 0 DE 0 AK 0 OR 0 PA

**Q: Write a few paragraphs about what you learned this week in the course. What interesting things did you discover in this assignment with simple queries?**

```
A:    The first thing I learned was how to properly use SELECT queries. It seemed a little tough at first but with a little practice the last couple problems in the assignment came pretty easily. My friend who worked in cyber security for over ten years gave me the best advice for this kind of stuff, just keep working at it and whatever seems hard will become easy. I'm not all the way to a full fledged data scientist but being able to grow is certainly a good feeling. The SELECT query is straight forward and there isn't any strange langauge(at least not yet!) for its use. I think if a layman looked at the code it would intimidate them but when you know something about how it works like the FROM just picks the table or ORDER BY helps you order things, the scariness goes away.

```
```
       I learned a few things from playing with the data. I thought looking at prices/costs in the data would be the most interesting place to start. For example I wanted to know what kind of components cost the most. The results clearly show that the wheels and the forks are the most expensive components availabe for a mountain bike. You have to go a little bit down the list to get to the most expensive shock which is over 20 rows down. I wanted to know how far down in price you would have to go from the peak to get out of the expert class of bikes and into the mid class of bikes. The cutoff appears to be just over $5000. I wonder if the quality of bike is actually noticable going from the most expensive mid bike to the least expensive expert bike?

```
```
       Going over the assignment data as well you can see that bikers enjoy fancy names for their products. Names like wreckoning and megatower belong to some of the most expensive bikes. These names could literally belong to anything but they sure sound cool. The same can even be said for individual components names. There could be some kind of rhyme or reason to these names but its not clear to me. 

```